### PR TITLE
Blocked by #61: Add recommendations for roles

### PR DESCRIPTION
**Pull request by jillr**
Draft of roles recommendations from ansible eng

cc @cidrblock

### CODE REVIEW

**Review by branic — APPROVED** [review]
LGTM

**Comment by ericzolf** [comment]
I don't agree with everything but I need a bit more time (and effort) to give a fair review.

**Review by NilashishC — APPROVED** [review]

File: roles/README.adoc
Comment (ganeshrn) [inline roles/README.adoc:119]: The argspec validation can be done in role argspec itself using conditional for different states. If validation is required at task level maybe it's an indication that role is getting complicated and logic should be moved to a separate role?

File: roles/README.adoc
Comment (ganeshrn) [inline roles/README.adoc:124]: Should it also support dry-run capability?

Comment (cidrblock) [reply roles/README.adoc:124]: Yes, good catch. check-mode/dry run should be supported.......... I'm assuming there will be cases were the endpoint system makes it impossible, I think it would be best to call this out as a limitation of the endpoint not an omission from the content

File: roles/README.adoc
Comment (jillr) [reply roles/README.adoc:404]: If anyone knows of any examples I would love to include them - I couldn't find any

**Review by priyamsahoo — APPROVED** [review]

**Comment by ericzolf** [comment]
As another side comment, we should wait for PR #61 to be merged before modifying the roles section too much, or @bontreger will have to redo his PR yet again, and that wouldn't be nice.

File: roles/README.adoc
Comment (mophahr) [inline roles/README.adoc:116]:
```suggestion
A role should have a single entry point. 
Requiring users to use `tasks_from: file.yml` is discouraged as this could cause important pre and post task files executed in `tasks/main.yml` to be skipped.
```

File: roles/README.adoc
Comment (mophahr) [inline roles/README.adoc:117]:
```suggestion
The `state` or `action` keyword should be passed to a role when a role has such a capability and be required when calling the role. 
The `tasks/main.yml` should load the appropriate task file based on the action or state requested by the user, rather than the consumer having to have intimate knowledge of the file structure layout within the role.
```

File: roles/README.adoc
Comment (mophahr) [inline roles/README.adoc:118]: I'm not convinced we need this. Maybe I also misnuderstand the rule. Could you please give an example?

File: roles/README.adoc
Comment (mophahr) [inline roles/README.adoc:119]: I have never done this in my roles, so I don't really understand what this is about. Couild you please be more specific and add an example?

File: roles/README.adoc
Comment (mophahr) [inline roles/README.adoc:122]: I believe we already have a rule for this. please double-check

File: roles/README.adoc
Comment (mophahr) [inline roles/README.adoc:123]:
```suggestion
All roles should be packaged within an Ansible collection.
```

This needs a bit more context.

File: roles/README.adoc
Comment (mophahr) [inline roles/README.adoc:124]:
```suggestion
Configuration management roles should be atomic, idempotent, and rollback when failures are encountered.
```

This is _huge_; especially the rollback part. I think we should discuss it in teh CoP meeting

File: roles/README.adoc
Comment (mophahr) [inline roles/README.adoc:125]:
```suggestion
When many roles within a collection are similar, consider using a dispatcher role as the single entry point to invoke a specific role within the collection. 
An example of such may be a collection that does system health checks and contains many roles, each representing an individual health check.
```

File: roles/README.adoc
Comment (mophahr) [inline roles/README.adoc:402]: why require `meta/main.yml`?

**Review by mophahr — CHANGES_REQUESTED** [review]
there's a few things that need clarification. See my individual comments.

Also, please have a look at the contribution guide (https://github.com/redhat-cop/automation-good-practices/blob/main/CONTRIBUTE.adoc)

**Review by ericzolf — CHANGES_REQUESTED** [review]
Sorry, prepared my review and forgot to send it, some duplicates/contradicts with @mophahr

File: roles/README.adoc
Comment (jillr) [reply roles/README.adoc:404]: @cidrblock @ganeshrn if we can't support the arg spec validation recommendations with documentation or functional/user-accessible sample code I'd like to remove them.  Any objections?

File: roles/README.adoc
Comment (tima) [reply roles/README.adoc:121]: Agreed. I don't feel we should be setting arbitrary limits like this.

File: roles/README.adoc
Comment (tima) [reply roles/README.adoc:124]: Yes I agree. There be demons here.

File: roles/README.adoc
Comment (tima) [reply roles/README.adoc:125]: This example and recommendation is a curious to me. Is this that common a scenario? If not I'd recommend we leave it out especially if a tool like ansible-lint cannot easily identify the pattern and make a recommendation. My overarching concern with these guidelines is that they will overwhelm a would-be Ansible content developer from getting their heads wrapped around all of these and discourages them from even attempting to create something. 

Are these health checks going to be potentially run as one off operations? If not, then I question the value creating individual roles. This seems counter to my long standing understanding of roles being coarse-grain use case driven units of Ansible content and not programming libraries. Individual health checks sounds too fine-grain unless there is a compelling reasons why each individual health check may be executed independently with their own distinct handlers, defaults and vars. When I hear "dispatcher role" and "many similar roles within one collection" I feel like we are talking about writing coding and I wonder if the person who wrote that content is a Java programmer.

File: roles/README.adoc
Comment (jillr) [reply roles/README.adoc:124]: @ericzolf What I understand here is that we should remove this recommendation from this PR/section in favor of the work happening in PR 42.  Please let me know if I misunderstand and I can add it back in.

File: roles/README.adoc
Comment (jillr) [reply roles/README.adoc:125]: @ssbarnea we need to be able to provide better documentation, examples, and rationale for suggestions like this.  I'm going to drop it from this PR, and it would need to be re-proposed as it's own change so that we don't drag out this PR if we want to keep it.

File: roles/README.adoc
Comment (jillr) [reply roles/README.adoc:495]: Why would stand alone roles not be appropriately licensed?

Comment (ericzolf) [reply roles/README.adoc:492]: Two main reasons:

1. we want to encourage distribution by collections, don't we? And licensing is a distribution topic, not so much a code topic, so (IMHO) the topic fits better in the collection section
2. the last bullet is explicitly relating to collections

Comment (jillr) [reply roles/README.adoc:492]: I can follow that logic.  This info is in #66 for collections so I'll remove it from this PR. Thanks.

File: roles/README.adoc
Comment (ericzolf) [inline roles/README.adoc:100]:
```suggestion
[%collapsible]
====
```

Comment (ericzolf) [reply roles/README.adoc:100]: That must have been the issue with the pipeline.

File: roles/README.adoc
Comment (ericzolf) [reply roles/README.adoc]: Good point, let's keep it here.

**Review by sean-m-sullivan — APPROVED** [review]
This looks good for a start, not sure what else to add.

File: roles/README.adoc
Comment (sean-m-sullivan) [reply roles/README.adoc]: Agreed, right now this is a pain point for controller configuration.

**Comment by ericzolf** [comment]
@mophahr will redo a review of the PR, it's easier as it's not that long.

@jillr you still need to address my suggestion at https://github.com/redhat-cop/automation-good-practices/pull/65/files#r915580222 because else the pipeline will fail.

File: roles/README.adoc
Comment (mophahr) [inline roles/README.adoc:384]: the bullet points here aren't rendered correctly. maybe this works:

```suggestion
At minimum include:

```

File: roles/README.adoc
Comment (mophahr) [inline roles/README.adoc:376]:
```suggestion
=== Create a meaningful README file for every role
```

File: roles/README.adoc
Comment (mophahr) [inline roles/README.adoc:388]: I'm not sure if I get what a "unit of automation is", could you please expand this a bit?

```suggestion
* The unit of automation the role provides
```

**Review by mophahr — CHANGES_REQUESTED** [review]
As announced by @ericzolf, I've reviewed the content. 

@jillr I have made some new small aesthetic suggestions you could apply. Once adressed, i think we can merge this.

@ericzolf could you please either approve or retract your "needs changes" on this PR?

**Comment by mophahr** [comment]
Thanks for the changes @jillr ! Approved from my side.

@ericzolf please remove/change your review.

**Review by mophahr — APPROVED** [review]

**Comment by ericzolf** [comment]
Approved in CoP meeting today. Again thanks for your patience and good work!

**Pull request merged**
